posthog/temporal/batch_exports/s3_batch_export.py

```diff
@@ -466,50 +491,142 @@ async def worker_shutdown_handler():
+            first_record_batch = cast_record_batch_json_columns(first_record_batch)
+            column_names = first_record_batch.column_names
+            column_names.pop(column_names.index("_inserted_at"))
+
+            schema = pa.schema(
+                # NOTE: For some reason, some batches set non-nullable fields as non-nullable, whereas other
+                # record batches have them as nullable.
+                # Until we figure it out, we set all fields to nullable. There are some fields we know
+                # are not nullable, but I'm opting for the more flexible option until we out why schemas differ
+                # between batches.
+                [field.with_nullable(True) for field in first_record_batch.select(column_names).schema]
+            )
+
+            writer = get_batch_export_writer(
+                inputs,
+                flush_callable=flush_to_s3,
+                max_bytes=settings.BATCH_EXPORT_S3_UPLOAD_CHUNK_SIZE_BYTES,
+                schema=schema,
+            )
+
+            async with writer.open_temporary_file():
                 rows_exported = get_rows_exported_metric()
                 bytes_exported = get_bytes_exported_metric()
 
-                async def flush_to_s3(last_uploaded_part_timestamp: str, last=False):
-                    logger.debug(
-                        "Uploading %s part %s containing %s records with size %s bytes",
-                        "last " if last else "",
-                        s3_upload.part_number + 1,
-                        local_results_file.records_since_last_reset,
-                        local_results_file.bytes_since_last_reset,
-                    )
+                for record_batch in record_iterator:
+                    record_batch = cast_record_batch_json_columns(record_batch)
 
-                    await s3_upload.upload_part(local_results_file)
-                    rows_exported.add(local_results_file.records_since_last_reset)
-                    bytes_exported.add(local_results_file.bytes_since_last_reset)
+                    await writer.write_record_batch(record_batch)
 
-                    activity.heartbeat(last_uploaded_part_timestamp, s3_upload.to_state())
+            await s3_upload.complete()
 
-                for record_batch in record_iterator:
-                    for record in record_batch.to_pylist():
-                        for json_column in ("properties", "person_properties", "set", "set_once"):
-                            if (json_str := record.get(json_column, None)) is not None:
-                                record[json_column] = json.loads(json_str)
+        return writer.records_total
 
-                        inserted_at = record.pop("_inserted_at")
 
-                        local_results_file.write_records_to_jsonl([record])
+def get_batch_export_writer(
+    inputs, flush_callable: FlushCallable, max_bytes: int, schema: pa.Schema | None = None
```

Comment:
```suggestion
    inputs: S3InsertInputs, flush_callable: FlushCallable, max_bytes: int, schema: pa.Schema | None = None
```

Reply:
Happy to commit. I didn't specify the type as eventually I want to make this a generic function for all destinations. But we can figure that out later once this is used everywhere.

Reply:
Resolved in latest commit. Wanted to push all commits in one go, so added this as the last one.

Reply:
> I didn't specify the type as eventually I want to make this a generic function for all destinations. But we can figure that out later once this is used everywhere.

Makes sense! I only noticed this because I was stepping through where `inputs` was passed and noticed the call stack ended here due to the lack of parameter type so figured it'd be worth adding. Can change it later if needed!